Sparse benchmarking utils

aocsa opened the pull request:
This is a benchmarking tooling to work with sparse tensors. To implement this, we extended PR `benchmarking util` [#38338](https://github.com/pytorch/pytorch/pull/38338) for sparse tensors.   In order to extend the proposed utility library the **FuzzedTensor** class was extended  by creating the new **FuzzedSparseTensor** class.  In addition two new operator classes were added, the `UnaryOpSparseFuzzer` and `BinaryOpSparseFuzzer`.

The class `FuzzedSparseTensor` adds new input parameters to the constructor: 
1. `sparse_dim`: The number of sparse dimensions in a sparse tensor.
2. `nnz`:   Number of non-zero elements in the sparse tensor. 
3. `density`: The density of the sparse tensor.
4. `coalesced`: As we know the sparse tensor format permits coalesced/uncoalesced sparse tensors.

and removes `probability_contiguous`, `max_allocation_bytes`, `roll_parameter`, `tensor_constructor` as they are dense-tensors related parameters.   

In addition, I've extended the `torch.utils.benchmark.examples` to work with the new classes `FuzzedSparseTensor`, `UnaryOpSparseFuzzer` and `BinaryOpSparseFuzzer`.

Hopefully, this tooling and these examples will help to make other benchmarks in other PRs. Looking forward to your thoughts and feedback. cc @robieta, @mruberry,  @ngimel

aocsa committed:
init sparse benchmarking proposal


update sparse bench


minor refactor and tests updated


minor update


minor update 2

aocsa committed:
minor fix

mruberry commented on torch/utils/benchmark/utils/sparse_fuzzer.py:
```diff
@@ -0,0 +1,114 @@
+        i = i.to(torch.long)
+
+        if not is_coalesced: 
+            v = torch.cat([v, torch.randn_like(v)], 0)
```
The nnz or the density vs. coalescing and non-coalescing behavior here seems odd.

When is_coalesced is false the number of elements is doubled!

When is_coalesced is true the number of elements is reduced in the coalescing process by an unclear amount.

aocsa replied:
Thanks, @mruberry I was trying to solve this, and I couldn't figure out  a clear and optimal solution. The main problem here is that the coalescing process will reduce the `nnz` to an unclear amount if the `torch.rand`  generates duplicates values in the process. This could be solved by using a method which generates random value without duplicates, however index values could represent and n-dimensional tensor, so to generate indices without duplicates is too expensive with this restriction.  What I was thinking is keep this behavior  because the probability to generate duplicates indices are low for most of the cases.  

In the other hand when `is_coalesced` is false, yes,  the number of elements is doubled but the number of indices represents the same amount of `nnz` for most of the cases. I mean virtually this represents the same  tensor with the same sparsity pattern.   Moreover, most of the sparse operation will use `coalesce()` method and what we want here is to get a sparse tensor with the same `nnz` even if this is coalesced or not. 

Let me know your thoughts.

mruberry replied:
I appreciate your point. OK, as long as we document this I think it's OK.

aocsa replied:
done

ngimel commented on torch/utils/benchmark/examples/sparse/op_benchmark.py:
```diff
@@ -0,0 +1,98 @@
+
+def run(n, stmt, fuzzer_cls):
+    float_iter = fuzzer_cls(seed=0, dtype=torch.float32).take(n)
+    int_iter = fuzzer_cls(seed=0, dtype=torch.float64).take(n)
```
this does not generate integer tensors, so don't call it `int_iter`

aocsa replied:
ok

ngimel reviewed (commented):
I feel like reusable parts of this PR are BinaryOpSparseFuzzer and UnaryOpSparseFuzzer, compare.py and fuzzer.py don't really add context compared to their non-sparse counterparts. For those, you probably want to specify densities and not nnzs though. 
If you want to use fuzzers to benchmark matmuls, you probably will need to create matmul sparse fuzzer. In the near term, since we probably are not going to focus on benchmarking unary and binary ops, working on the infrastructure for matmuls is more valuable.

aocsa committed:
minor update

aocsa committed:
minor updates

aocsa commented on torch/utils/benchmark/op_fuzzers/sparse_binary.py:
```diff
@@ -0,0 +1,106 @@
+                ],
+                FuzzedParameter(
+                    name="nnz",
+                    distribution={5: 0.4, 10: 0.3, 15: 0.3},
```
Sure, density makes more sense.

aocsa commented on torch/utils/benchmark/utils/sparse_fuzzer.py:
```diff
@@ -0,0 +1,114 @@
+                The number of sparse dimensions in a sparse tensor.
+            nnz: 
+                Number of non-zero elements in the sparse tensor. 
+            density: The `nnz` value will be truncated to this value.
```
I see, now fuzzers only works with density parameter

aocsa commented on torch/utils/benchmark/examples/sparse/compare.py:
```diff
@@ -0,0 +1,126 @@
+"""Example of Timer and Compare APIs:
```
fuzzer.py is just an and example about how to use the new FuzzedSparseTensor. In the other hand `compare.py` is an example of how to bench matmul operation over sparse matrices. A more complete benchmark reference for matmul operations was submitted previously in a different PR  https://github.com/pytorch/pytorch/blob/master/benchmarks/sparse/matmul_dlmc_bench.py. This benchmark uses a popular and real sparse matrix dataset https://github.com/google-research/google-research/tree/master/sgk.

aocsa committed:
adding documentation and minor update

aocsa commented:
I addressed review comments. 

> I feel like reusable parts of this PR are BinaryOpSparseFuzzer and UnaryOpSparseFuzzer, compare.py and fuzzer.py don't really add context compared to their non-sparse counterparts. For those, you probably want to specify densities and not nnzs though.

> If you want to use fuzzers to benchmark matmuls, you probably will need to create matmul sparse fuzzer. In the near term, since we probably are not going to focus on benchmarking unary and binary ops, working on the infrastructure for matmuls is more valuable.

Regarding this request, I previously worked doing sparse matmul [benchmarks](https://github.com/pytorch/pytorch/blob/master/benchmarks/sparse/matmul_dlmc_bench.py) using a real sparse matrix [dataset](https://github.com/google-research/google-research/tree/master/sgk).   Would you think necessary to use a synthetic dataset implementing a matmul sparse fuzzer instead of using a real dataset? 

@ngimel, @mruberry - looking forward your comments.

mruberry commented:
> I addressed review comments.
> 
> > I feel like reusable parts of this PR are BinaryOpSparseFuzzer and UnaryOpSparseFuzzer, compare.py and fuzzer.py don't really add context compared to their non-sparse counterparts. For those, you probably want to specify densities and not nnzs though.
> 
> > If you want to use fuzzers to benchmark matmuls, you probably will need to create matmul sparse fuzzer. In the near term, since we probably are not going to focus on benchmarking unary and binary ops, working on the infrastructure for matmuls is more valuable.
> 
> Regarding this request, I previously worked doing sparse matmul [benchmarks](https://github.com/pytorch/pytorch/blob/master/benchmarks/sparse/matmul_dlmc_bench.py) using a real sparse matrix [dataset](https://github.com/google-research/google-research/tree/master/sgk). Would you think necessary to use a synthetic dataset implementing a matmul sparse fuzzer instead of using a real dataset?
> 
> @ngimel, @mruberry - looking forward your comments.

A real dataset would be great!

mruberry reviewed (approved):
Cool, thanks @aocsa!
